internal/server/task.go

```diff
@@ -0,0 +1,143 @@
+package server
+
+import (
+	"fmt"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/ministryofjustice/opg-go-common/template"
+	"github.com/ministryofjustice/opg-sirius-lpa-frontend/internal/sirius"
+	"golang.org/x/sync/errgroup"
+)
+
+type TaskClient interface {
+	CreateTask(ctx sirius.Context, task sirius.Task) error
+	TaskTypes(ctx sirius.Context) ([]string, error)
+	Teams(ctx sirius.Context) ([]sirius.Team, error)
+	Case(ctx sirius.Context, id int) (sirius.Case, error)
+}
+
+type taskData struct {
+	XSRFToken string
+	Today     string
+	Entity    string
+	Success   bool
+	Errors    sirius.ValidationErrors
+
+	TaskTypes        []string
+	Teams            []sirius.Team
+	Task             sirius.Task
+	AssignTo         string
+	AssigneeUserName string
+}
+
+func Task(client TaskClient, tmpl template.Template) Handler {
+	return func(w http.ResponseWriter, r *http.Request) error {
+		caseID, err := strconv.Atoi(r.FormValue("id"))
+		if err != nil {
+			return err
+		}
+
+		ctx := getContext(r)
+		data := taskData{
+			XSRFToken: ctx.XSRFToken,
+			Today:     time.Now().Format("2006-01-02"),
+		}
+
+		group, groupCtx := errgroup.WithContext(ctx.Context)
+
+		group.Go(func() error {
+			taskTypes, err := client.TaskTypes(ctx.With(groupCtx))
+			if err != nil {
+				return err
+			}
+
+			data.TaskTypes = taskTypes
+			return nil
+		})
+
+		group.Go(func() error {
+			teams, err := client.Teams(ctx.With(groupCtx))
+			if err != nil {
+				return err
+			}
+
+			data.Teams = teams
+			return nil
+		})
+
+		group.Go(func() error {
+			caseitem, err := client.Case(ctx.With(groupCtx), caseID)
+			if err != nil {
+				return err
+			}
+			data.Entity = fmt.Sprintf("%s %s", caseitem.CaseType, caseitem.UID)
+			return nil
+		})
+
+		if err := group.Wait(); err != nil {
+			return err
+		}
+
+		if r.Method == http.MethodPost {
+			task := sirius.Task{
+				CaseID:      caseID,
+				Type:        r.FormValue("type"),
+				DueDate:     formatSiriusDate(r.FormValue("dueDate")),
+				Name:        r.FormValue("name"),
+				Description: r.FormValue("description"),
+			}
+			assignTo := r.FormValue("assignTo")
+
+			switch assignTo {
+			case "user":
+				parts := strings.SplitN(r.FormValue("assigneeUser"), ":", 2)
```

Comment:
I needed a way of getting the name of the selected assignee back to rerender the form, this seemed better than adding in a request to get the assignee :shrug: